### PR TITLE
Vertically scale formula bar

### PR DESCRIPTION
**Pull request by naterush**
# Description

This vertically scales formula bar. The one final thing we could do here is automatically add new lines when you do, but this is kinda annoying (we have to save height or something?) so I'd be down to ignore it.

# Testing

Write some formulas yo.

# Documentation

No.

### CODE REVIEW

**Comment by aarondr77** [comment]
In order to support the kinds of structure we think will be helpful, it would be nice to add support for `tab`. Otherwise, I use spaces in the formula bar that end up doing this:

Screenshot 2023-04-11 at 10 40 41 AM

Screenshot 2023-04-11 at 10 40 12 AM

**Comment by aarondr77** [comment]
In the work I'm doing with some users right now, we would be submitting this formula to check the result and then wanting to update it right away, it would be nice if the formula bar saved its height. Some ideas:

1. The next time I open the formula, it goes back to the same height
2. I can resize the formula bar when it's closed or open and the height just persists. 
3. Add a keyboard shortcut with a hint in the tool that tells me how to make it larger without having to go drag it

**Comment by aarondr77** [comment]
There's some bug where if I try to open the formula bar on a column header Mito freezes

https://user-images.githubusercontent.com/18709905/231200892-11c14b2f-3641-46d0-85de-072463142664.mov

**Comment by aarondr77** [comment]
We should just add a max height to the text area

https://user-images.githubusercontent.com/18709905/231201173-ee7e9fc0-48ab-42b4-90e0-d46e7f90af62.mov

**Comment by aarondr77** [comment]
While we are here, in Excel, if you single click on the formula bar it lets you start editing. Can we do that? I saw someone really struggle yesterday to get it to open.

File: mitosheet/src/components/endo/celleditor/cellEditorUtils.tsx [hunk 1]
```diff
@@ -188,12 +188,27 @@ export const getStartingFormula = (
...
+
+
+    let found = false;
+    lowercaseColumnHeaders.forEach(ch => {
+        const lastIndexOf = lowercaseFormula.lastIndexOf(ch);
+        if (lastIndexOf !== -1) {
+            const remainingString = lowercaseFormula.substring(lastIndexOf);
+            // Check if this is an index label (TODO: is this performant enough?)
+            sheetData.index.forEach(indexLabel => {
+                if (remainingString.endsWith(getDisplayColumnHeader(indexLabel).toLowerCase())) {
```
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/cellEditorUtils.tsx:204]: Should be checking equals, not ends with!

File: mitosheet/src/components/endo/celleditor/CellEditorDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    editorState: EditorState,
+    selectionStart: number | null | undefined,
+    cellEditorError: string | undefined,
+    loading: boolean,
+): DisplayedDropdownType | undefined => {
+
+    const fullFormula = getFullFormula(editorState.formula, editorState.pendingSelections, sheetData);
+    const endsInReference = getFormulaEndsInReference(fullFormula, sheetData);
+
+    console.log("ENDS IN REFERNECE")
```
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditorDropdown.tsx:45]: Remove

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx [hunk 3]
```diff
@@ -47,16 +55,20 @@ const CellEditor = (props: {
...
-    const cellEditorInputRef = useRef<HTMLInputElement | null>(null);
+    const fullFormula = getFullFormula(props.editorState.formula, props.editorState.pendingSelections, props.sheetData);
+
+    const cellEditorInputRef = useRef<HTMLInputElement | HTMLTextAreaElement | null>(null);
 
     const [selectedSuggestionIndex, setSavedSelectedSuggestionIndex] = useState(-1);
     const [loading, setLoading] = useState(false);
     const [cellEditorError, setCellEditorError] = useState<string | undefined>(undefined);
+    const [selectionRangeToSet, setSelectionRangeToSet] = useState<number|undefined>(undefined) // Allows us to place the cursor at a specific location
+    const [textAreaHeight, setTextAreaHeight] = useState(() => getDefaultTextAreaHeight(fullFormula)); // Default to 2 lines, 11px tall
```
Comment (naterush) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:66]: Remove comment

**Review by naterush — COMMENTED** [review]
Cleanup!

**Comment by naterush** [comment]
Ok, there are two things I want to do still:
1. Fix up the column header editing (still just disappears, not sure why)
2. Make it so the suggested column headers work in the middle of the formula, not just at the end (like the getDocumentationFuction).

**Comment by aarondr77** [comment]
When using the formula bar, the documentation covers up such a huge amount of the screen. I wanted to reference a column header, but I can't click on it and I also cannot use the arrow keys to reference it. 

An easy partial solution is to keep the documentation at the starter width instead of being the max width documentation, so that it covers up less of the grid. 

Screenshot 2023-04-12 at 2 14 44 PM

**Comment by aarondr77** [comment]
Turn off spell check 

Screenshot 2023-04-12 at 2 18 54 PM

**Comment by aarondr77** [comment]
Can we make the text area not have this weird extra space on the bottom?

Screenshot 2023-04-12 at 2 20 12 PM

**Comment by aarondr77** [comment]
When we click into it, I think it would be more natural if the cursor went to the end of the input instead of the beginning. 

https://user-images.githubusercontent.com/18709905/231549736-7943040e-5dc6-4082-8281-96bc8578fe86.mov

**Comment by aarondr77** [comment]
Can we get rid of these tabs in the generated code?

Screenshot 2023-04-12 at 2 23 52 PM

Screenshot 2023-04-12 at 2 23 28 PM

**Comment by aarondr77** [comment]
The fact that the formula bar cell editor opens back up to multi line in the above image is awesome!!!

**Comment by naterush** [comment]
All of your above comments have been addressed!

File: mitosheet/src/components/endo/celleditor/CellEditor.tsx [hunk 4]
```diff
@@ -141,7 +141,9 @@ const CellEditor = (props: {
         setTextAreaHeight(newHeight)
     }, [], 'cell-editor-input')
 
-    if (columnID === undefined || columnHeader === undefined || indexLabel === undefined) {
+    console.log("HERE", columnID, columnHeader, indexLabel)
```
Comment (aarondr77) [inline mitosheet/src/components/endo/celleditor/CellEditor.tsx:147]: remove

**Comment by aarondr77** [comment]
Linking here for visibility, but we should not do now: https://github.com/mito-ds/monorepo/issues/699

File: mitosheet/mitosheet/parser.py [hunk 5]
```diff
@@ -732,6 +732,25 @@ def get_header_header_selection_code(column_header_one: ColumnHeader, column_hea
...
+    
+def replace_newlines_and_tabs(
+        formula: str,
+    ) -> str:
+    """
+    Removes all newlines and tabs that aren't in column headers
+    """
+    string_matches = get_string_matches(formula)
+
+    def replace_newlines_and_tabsinternal(match):
```
Comment (aarondr77) [inline mitosheet/mitosheet/parser.py:745]: replace_newlines_and_tabsinternal -> replace_newlines_and_tabs_internal

**Comment by aarondr77** [comment]
If I try to add a tab to the beginning of the formula, it jumps to the end

https://user-images.githubusercontent.com/18709905/231563085-dad8ae57-ab66-463d-867d-9f84f8f0e62a.mov

**Comment by aarondr77** [comment]
If the formula has a tab at the beginning, it errors. 

Screenshot 2023-04-12 at 3 23 39 PM

**Comment by aarondr77** [comment]
Got into this weird state where clicking in the grid does not append anything to formula. 

https://user-images.githubusercontent.com/18709905/231564320-d7a8b184-a3e6-4b7b-97ff-3dc7f0f7486a.mov

File: mitosheet/mitosheet/parser.py [hunk 6]
```diff
@@ -818,9 +837,9 @@ def parse_formula(
         df_name
     )
 
-    code_with_functions, functions = replace_functions(
-        code_with_column_headers,
-    )
+    code_without_newlines = replace_newlines_and_tabs(code_with_column_headers)
```
Comment (aarondr77) [inline mitosheet/mitosheet/parser.py:840]: code_without_newlines -> code_without_newlines_or_tabs

**Review by aarondr77 — COMMENTED** [review]
Sweet! This is working really well. Feels pretty smooth!! LGTM once after fixing these last few things up. 

I tested this on Mac/Safari, we should make sure to test it on other browsers before releasing

**Comment by naterush** [comment]
Ok, once tests pass we are ready to merge.

**Pull request merged**